### PR TITLE
Add unit tests for UDPServer

### PR DESCRIPTION
**Pull request by pafuent**
This PR aims to help "fix" #43440

### CODE REVIEW

**Review by radiantgurl — APPROVED** [review]

File: tests/core/io/test_udp_server.h [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+TEST_CASE("[UDPServer] Accept a connection and receive/send data") {
+	Ref<UDPServer> server = create_server(localhost, port);
+	Ref<PacketPeerUDP> client = create_client(localhost, port);
+
+	// Sending data from client to server
+	String hello_world = "Hello World!";
+	CHECK_EQ(Error::OK, client->put_var(hello_world));
+
+	// Required to get the connection properly stablished
```
Comment (Calinou) [inline tests/core/io/test_udp_server.h:97]:
```suggestion
	// Required to get the connection properly established
```

File: tests/core/io/test_udp_server.h [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+	Variant hello_world_received;
+	Ref<PacketPeerUDP> client_from_server = accept_connection(server);
+	CHECK_EQ(Error::OK, client_from_server->get_var(hello_world_received));
+	CHECK_EQ(hello_world, String(hello_world_received));
+
+	// Sending data from server to client
+	Variant pi = 3.1415;
+	CHECK_EQ(Error::OK, client_from_server->put_var(pi));
+
+	// Required to get the connection properly stablished
```
Comment (Calinou) [inline tests/core/io/test_udp_server.h:109]:
```suggestion
	// Required to get the connection properly established
```

File: tests/core/io/test_udp_server.h [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+		Ref<PacketPeerUDP> c = create_client(localhost, port);
+
+		// Sending data from client to server
+		String hello_client = "Hello " + itos(i);
+		CHECK_EQ(Error::OK, c->put_var(hello_client));
+
+		clients.push_back(c);
+	}
+
+	// Required to get the connection properly stablished
```
Comment (Calinou) [inline tests/core/io/test_udp_server.h:138]:
```suggestion
	// Required to get the connection properly established
```

File: tests/core/io/test_udp_server.h [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+		Variant hello_world_received;
+		CHECK_EQ(Error::OK, cfs->get_var(hello_world_received));
+		CHECK_EQ("Hello " + itos(i), String(hello_world_received));
+
+		// Sending data from server to client
+		Variant pi = 3.1415 + i;
+		CHECK_EQ(Error::OK, cfs->put_var(pi));
+	}
+
+	// Required to get the connection properly stablished
```
Comment (Calinou) [inline tests/core/io/test_udp_server.h:153]:
```suggestion
	// Required to get the connection properly established
```

File: tests/core/io/test_udp_server.h [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+	CHECK_EQ(hello_world, String(hello_world_received));
+
+	client->close();
+	server->stop();
+	CHECK_FALSE(server->is_listening());
+
+	Ref<PacketPeerUDP> new_client = create_client(localhost, port);
+	CHECK_EQ(Error::OK, new_client->put_var(hello_world));
+
+	// Required to get the connection properly stablished
```
Comment (Calinou) [inline tests/core/io/test_udp_server.h:196]:
```suggestion
	// Required to get the connection properly established
```

File: tests/core/io/test_udp_server.h [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+
+TEST_CASE("[UDPServer] Should disconnect client") {
+	Ref<UDPServer> server = create_server(localhost, port);
+	Ref<PacketPeerUDP> client = create_client(localhost, port);
+
+	// Sending data from client to server
+	String hello_world = "Hello World!";
+	CHECK_EQ(Error::OK, client->put_var(hello_world));
+
+	// Required to get the connection properly stablished
```
Comment (Calinou) [inline tests/core/io/test_udp_server.h:216]:
```suggestion
	// Required to get the connection properly established
```

File: tests/core/io/test_udp_server.h [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+	Ref<PacketPeerUDP> client_from_server = accept_connection(server);
+	CHECK_EQ(Error::OK, client_from_server->get_var(hello_world_received));
+	CHECK_EQ(hello_world, String(hello_world_received));
+
+	server->stop();
+	CHECK_FALSE(server->is_listening());
+	CHECK_FALSE(client_from_server->is_bound());
+	CHECK_FALSE(client_from_server->is_socket_connected());
+
+	// Required to get the connection properly stablished
```
Comment (Calinou) [inline tests/core/io/test_udp_server.h:229]:
```suggestion
	// Required to get the connection properly established
```

File: tests/core/io/test_udp_server.h [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+	OS::get_singleton()->delay_usec(500000);
+
+	// Sending data from client to server
+	CHECK_EQ(Error::OK, client->put_var(hello_world));
+
+	// Required to get the connection properly stablished
```
Comment (Calinou) [inline tests/core/io/test_udp_server.h:235]:
```suggestion
	// Required to get the connection properly established
```

File: tests/core/io/test_udp_server.h [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+		Ref<PacketPeerUDP> c = create_client(localhost, port);
+
+		// Sending data from client to server
+		String hello_client = "Hello " + itos(i);
+		CHECK_EQ(Error::OK, c->put_var(hello_client));
+
+		clients.push_back(c);
+	}
+
+	// Required to get the connection properly stablished
```
Comment (Calinou) [inline tests/core/io/test_udp_server.h:260]:
```suggestion
	// Required to get the connection properly established
```

File: tests/core/io/test_udp_server.h (hunk 1 above)
Comment (AThousandShips) [reply tests/core/io/test_udp_server.h:97]:
```suggestion
	// Required to get the connection properly established.
```
Period at end as well

File: tests/core/io/test_udp_server.h [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+TEST_CASE("[UDPServer] Accept a connection and receive/send data") {
+	Ref<UDPServer> server = create_server(localhost, port);
+	Ref<PacketPeerUDP> client = create_client(localhost, port);
+
+	// Sending data from client to server.
+	String hello_world = "Hello World!";
+	CHECK_EQ(Error::OK, client->put_var(hello_world));
+
+	// Required to get the connection properly established.
+	OS::get_singleton()->delay_usec(500000);
```
Comment (Faless) [inline tests/core/io/test_udp_server.h:98]: It's always recommended to `poll` while waiting

**Review by Calinou — CHANGES_REQUESTED** [review]

**Comment by pafuent** [comment]
@Calinou I addressed all your comments, Could you please review again this PR?

**Comment by pafuent** [comment]
Friendly remainder

**Comment by Faless** [comment]
> Friendly remainder

As mentioned in https://github.com/godotengine/godot/pull/97262#discussion_r1805889374 :

> The problem here is that you keep blocking for a long time and only poll once, while you should only block for brief periods of time and regularly poll (i.e. **multiple times**), or you risk of having the same error on other OS configurations.

**Comment by pafuent** [comment]
@Calinou I addressed all your comments, Could you please review if I addressed them correctly?

**Review by Calinou — APPROVED** [review]
Looks good to me.

File: tests/core/io/test_udp_server.h [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+	// Required to get the connection properly established.
+	const uint64_t time = OS::get_singleton()->get_ticks_usec();
+	while (!server->is_connection_available() && (OS::get_singleton()->get_ticks_usec() - time) < MAX_WAIT_USEC) {
+		server->poll();
+		OS::get_singleton()->delay_usec(SLEEP_DURATION);
+	}
+
+	CHECK_EQ(server->poll(), Error::OK);
+	REQUIRE(server->is_connection_available());
+	Ref<PacketPeerUDP> client_from_server = server->take_connection();
```
Comment (AThousandShips) [inline tests/core/io/test_udp_server.h:80]:
```suggestion
Ref<PacketPeerUDP> accept_connection(Ref<UDPServer> &p_server) {
	// Required to get the connection properly established.
	const uint64_t time = OS::get_singleton()->get_ticks_usec();
	while (!p_server->is_connection_available() && (OS::get_singleton()->get_ticks_usec() - time) < MAX_WAIT_USEC) {
		p_server->poll();
		OS::get_singleton()->delay_usec(SLEEP_DURATION);
	}

	CHECK_EQ(p_server->poll(), Error::OK);
	REQUIRE(p_server->is_connection_available());
	Ref<PacketPeerUDP> client_from_server = p_server->take_connection();
```

**Review by AThousandShips — DISMISSED** [review]

**Review by Faless — APPROVED** [review]
As discussed with other members, I'm approving the PR, and will make the necessary changes in a separate PR.

**Pull request merged**